diff --git a/tensorrt_llm/runtime/multimodal_model_runner.py b/tensorrt_llm/runtime/multimodal_model_runner.py
index 5f9da08c..cd36bb2e 100644
--- a/tensorrt_llm/runtime/multimodal_model_runner.py
+++ b/tensorrt_llm/runtime/multimodal_model_runner.py
@@ -685,6 +685,30 @@ class MultimodalModelRunner:
             self.vision_model.get_img_features(image)
             return
 
+        if self.model_type == "phi-4-multimodal":
+            model = AutoModelForCausalLM.from_pretrained(
+                self.args.hf_model_dir,
+                torch_dtype=torch.float16,
+                trust_remote_code=True,
+                device_map='cpu')
+            vision_model = model.model.embed_tokens_extend.image_embed
+            self.vision_model = vision_model.to(self.device, torch.float16).eval()
+
+            # Test run vision_model.get_img_features to pre-allocate memory for flash attention
+            image = self.processor(text="<|image_1|>",
+                                   images=Image.new('RGB', [10, 10]),
+                                   return_tensors="pt")
+            # processor: ['input_ids', 'input_image_embeds', 'image_sizes', 'image_attention_mask', 'input_audio_embeds', 'audio_embed_sizes', 'audio_attention_mask', 'attention_mask', 'input_mode']
+            # image_processor ----> dict_keys(['input_image_embeds', 'image_sizes', 'image_attention_mask', 'num_img_tokens'])
+            # audio_processor ----> dict_keys(['input_audio_embeds', 'audio_embed_sizes'])
+            image = image['input_image_embeds']
+            image = image.flatten(0, 1)
+            image = torch.rand(image.shape,
+                               dtype=str_dtype_to_torch(self.vision_precision),
+                               device=self.device)
+            self.vision_model.get_img_features(image)
+            return
+
         if self.cpp_e2e:
             logger.info(
                 "Using C++ runtime for both visual engine and LLM decoder, skip loading visual engine in Python runtime."
@@ -725,14 +749,36 @@ class MultimodalModelRunner:
 
     def init_audio_encoder(self):
         assert self.model_type == "phi-4-multimodal"
-        audio_encoder_path = os.path.join(self.audio_engine_dir,
-                                          self.args.audio_engine_name)
-        logger.info(f'Loading engine from {audio_encoder_path}')
-        with open(audio_encoder_path, 'rb') as f:
-            engine_buffer = f.read()
-        logger.info(f'Creating session from engine {audio_encoder_path}')
-        self.audio_encoder_session = Session.from_serialized_engine(
-            engine_buffer)
+        # Older code that worked with Eager Attention.
+        # audio_encoder_path = os.path.join(self.audio_engine_dir,
+        #                                   self.args.audio_engine_name)
+        # logger.info(f'Loading engine from {audio_encoder_path}')
+        # with open(audio_encoder_path, 'rb') as f:
+        #     engine_buffer = f.read()
+        # logger.info(f'Creating session from engine {audio_encoder_path}')
+        # self.audio_encoder_session = Session.from_serialized_engine(
+        #     engine_buffer)
+
+        model = AutoModelForCausalLM.from_pretrained(
+            self.args.hf_model_dir,
+            torch_dtype=torch.float16,
+            trust_remote_code=True,
+            device_map='cpu')
+
+        audio_model = model.model.embed_tokens_extend.audio_embed
+        self.audio_model = audio_model.to(self.device, torch.float16).eval()
+
+        # Do we need this for new code? What's the significance of this?
+        # # Test run vision_model.get_img_features to pre-allocate memory for flash attention
+        # image = self.processor(text="<|image_1|>",
+        #                         images=Image.new('RGB', [10, 10]),
+        #                         return_tensors="pt")['pixel_values']
+        # image = image.flatten(0, 1)
+        # image = torch.rand(image.shape,
+        #                     dtype=str_dtype_to_torch(self.vision_precision),
+        #                     device=self.device)
+        # self.vision_model.get_img_features(image)
+        # return
 
     def init_llm(self):
         if self.decoder_llm:
@@ -915,6 +961,10 @@ class MultimodalModelRunner:
                     image).reshape(1, image.shape[0], -1,
                                    self.vision_model.image_dim_out)
                 visual_atts = None
+            elif self.model_type == "phi-4-multimodal":
+                visual_features = self.vision_model.get_img_features(
+                            img_embeds, attention_mask)
+                visual_features = self.vision_model.img_projection(visual_features)
             else:
                 if self.cpp_e2e:
                     # If using E2E C++ runtime, visual_features will not be computed here in Python runtime.
@@ -941,6 +991,7 @@ class MultimodalModelRunner:
                     model_runner_input = torch.vsplit(
                         model_runner_input, model_runner_input.shape[0])
                 else:
+                    # @B: Previously, TRT engines would go here.
                     visual_features, visual_atts = self.get_visual_features(
                         model_runner_input, other_vision_inputs)
                     model_runner_input = None
@@ -1528,6 +1579,7 @@ class MultimodalModelRunner:
         return embeds
 
     def get_visual_features(self, image, other_vision_inputs):
+        assert False, "This should not be called. Video encoder must be called directly for Phi4MM."
         visual_features = {
             self.vision_input_names[0]:
             image.to(str_dtype_to_torch(self.vision_precision)),
@@ -1569,6 +1621,7 @@ class MultimodalModelRunner:
         return image_embeds, image_atts
 
     def get_audio_features(self, audio, other_audio_inputs):
+        assert False, "This should not be called. Audio encoder must be called directly for Phi4MM."
         audio_features = {
             self.audio_input_names[0]:
             audio.to(str_dtype_to_torch(self.audio_precision)),
diff --git a/tensorrt_llm/tools/multimodal_builder.py b/tensorrt_llm/tools/multimodal_builder.py
index e83611e5..22ccc464 100644
--- a/tensorrt_llm/tools/multimodal_builder.py
+++ b/tensorrt_llm/tools/multimodal_builder.py
@@ -1036,9 +1036,9 @@ def build_phi4mm_engine(args):
     os.makedirs(os.path.join(args.output_dir, "audio"), exist_ok=True)
     to_json_file(config_args, args.output_dir + "vision/config.json")
     to_json_file(config_args, args.output_dir + "audio/config.json")
-    
     return
 
+    # Keeping this part of the code as is in case there are fixes in ONNX/Pytorch in the future.
     processor = AutoProcessor.from_pretrained(args.model_path,
                                               trust_remote_code=True)
     raw_image = Image.new('RGB', [10, 10])  # dummy image
