diff --git a/src/include/non_abi/device/pt-to-pt/utils_device.h b/src/include/non_abi/device/pt-to-pt/utils_device.h
index 8342ebe06..01667c858 100644
--- a/src/include/non_abi/device/pt-to-pt/utils_device.h
+++ b/src/include/non_abi/device/pt-to-pt/utils_device.h
@@ -27,6 +27,8 @@
     *x = ((*(x)&0xFF000000) >> 24 | (*(x)&0x00FF0000) >> 8 | (*(x)&0x0000FF00) << 8 | \
           (*(x)&0x000000FF) << 24)
 
+#ifdef __CUDA_ARCH__
+
 #ifdef NVSHMEMI_COMM_DEVICE_UTILS_USE_PTX
 
 __device__ static inline uint64_t BSWAP64(uint64_t x) {
@@ -99,6 +101,8 @@ __device__ static inline uint16_t BSWAP16(uint16_t x) {
 
 #endif /* NVSHMEMI_COMM_DEVICE_UTILS_USE_PTX */
 
+#endif /* __CUDA_ARCH__ */
+
 #define HTOBE64(x) BSWAP64(x)
 #define HTOBE32(x) BSWAP32(x)
 #define HTOBE16(x) BSWAP16(x)
