diff --git a/block_sparse.py b/block_sparse.py
index a69006bcf8d6dd38e380e19fb8c3a22ace3a0545..3e1996e3fa4a834245b5207b8507d02a2bc36201 100644
--- a/block_sparse.py
+++ b/block_sparse.py
@@ -1,3 +1,4 @@
+# Copyright (c) 2026, NVIDIA CORPORATION. All rights reserved.
 # Copyright (c) 2026 by FlashInfer team.
 #
 # Licensed under the Apache License, Version 2.0 (the "License");
diff --git a/decode.py b/decode.py
index d08f4c9a014cef00e56d47e030807094c383962c..2334723b4d84e0d0d3ed7bb572bc839fd38fec22 100644
--- a/decode.py
+++ b/decode.py
@@ -1,3 +1,4 @@
+# Copyright (c) 2026, NVIDIA CORPORATION. All rights reserved.
 # Copyright (c) 2026 by FlashInfer team.
 #
 # Licensed under the Apache License, Version 2.0 (the "License");
diff --git a/mla_decode.py b/mla_decode.py
index 827f9413a7f2a1f2072b639fef428fb7b17de60b..dfd77328409d6f4ce9287cfb2203884901a476b6 100644
--- a/mla_decode.py
+++ b/mla_decode.py
@@ -1,3 +1,4 @@
+# Copyright (c) 2026, NVIDIA CORPORATION. All rights reserved.
 # Copyright (c) 2026 by FlashInfer team.
 #
 # Licensed under the Apache License, Version 2.0 (the "License");
